Youtube search filters

Bluesir9 opened the pull request:
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I did test the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [x] I agree to ASAP create a PULL request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) for making in compatible when I changed the api.

theScrabi commented:
You can find a description on how to test the changes on the frontend here: https://teamnewpipe.github.io/documentation/04_Run_changes_in_App/

Bluesir9 commented:
@theScrabi Have tested the changes on the app and made necessary changes.

theScrabi commented:
All right thx :)

TheAssassin commented on extractor/src/main/java/org/schabi/newpipe/extractor/utils/Base64.java:
```diff
@@ -0,0 +1,725 @@
+ * omitted on Android, so we cannot use this class on Android.
+ *
+ * Hence below implementation was picked from AOSP here ->
+ * https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/java/android/util/Base64.java
```
Depends on the license. In case of the Apache 2.0 license, I think you must:

* copy the original license header (contents) into this source file
* include a copy of the license somewhere in the repository
* highlight changes in the code "prominently" (i.e., in a comment)
* in case AOSP contains a `NOTICE` file, there's some more rules

The Apache 2.0 license is quite complex to get right. You should think about using a different implementation.

Bluesir9 replied:
I am not well versed with licenses so using this seemed like a no-brainer. Their is no straight-forward way of performing Base64 encoding when restricted to Java 7. There is the Apache commons library which provides this functionality but adding a library dependency for this singular use case seemed excessive.

TobiGr replied:
@Bluesir9 This is definetely a good intention and keeps the extractor small.
@TheAssassin I just went through the [NOTICE file](https://github.com/aosp-mirror/platform_frameworks_base/blob/master/NOTICE) and couldn't find any specific paragraph which matches the base64.java file, because it was written for/by the AOSP. Additionally, @Bluesir9  didn't change the code. So I guess, only these two steps are required:

> - copy the original license header (contents) into this source file
> - include a copy of the license somewhere in the repository

Bluesir9 replied:
@TobiGr If this is all then I will go ahead and make the change in sometime.

Bluesir9 replied:
@TobiGr Uh I have added the original license header in the Base64 class file as per your first point. What license needs to be copied into the repo for the second point?

TobiGr replied:
The license starts [here](https://github.com/aosp-mirror/platform_frameworks_base/blob/47d385a5b3c76111dedef11cb814486dacd77b8c/NOTICE#L108). I'd suggest to create a new 3rd-party-licenses directory which contains all the licenses apart from GPL3. @TheAssassin What do you think?

TheAssassin replied:
I guess that should work.

Bluesir9 commented on extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/linkHandler/YoutubeSearchQueryHandlerFactory.java:
```diff
@@ -2,19 +2,83 @@
+    }
+
+    public enum Filter {
+        All(FilterType.Content,(byte)0),
```
Ran automatic formatting to existing and new code.

Bluesir9 commented on extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/search/YoutubeSearchExtractorChannelOnlyTest.java:
```diff
@@ -8,28 +8,35 @@
 
 public class YoutubeSearchExtractorChannelOnlyTest extends YoutubeSearchExtractorBaseTest {
 
+    private static final String SEARCH_QUERY = "pewdiepie";
```
My changes affected this test so I changed it accordingly while putting the repeated stuff in a single place. What the tests were testing remains the same.

Bluesir9 commented on extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/linkHandler/YoutubeSearchQueryHandlerFactory.java:
```diff
@@ -23,31 +87,118 @@ public static YoutubeSearchQueryHandlerFactory getInstance() {
+        try {
+            sorter = Sorter.valueOf(filter);
+        } catch (IllegalArgumentException e) {
+            e.printStackTrace();
```
I have moved the exceptions all the way up to throw a ParsingException with the applicable exception. I am not sure if I can test this on a device though. Let me know if there is something you would like me to confirm/change for this.

theScrabi replied:
@Bluesir9 any news on this?

Bluesir9 replied:
@theScrabi This was addressed in prior commits. Your comment is linked to old code.

theScrabi replied:
I see :)

Bluesir9 commented:
@TobiGr For whatever reason "Feature" filters were not working with the changes I had done last time around, so in the interest of time I removed them last minute. I have implemented them now in the last couple of commits and its working fine. Have also added tests for the same.

TheAssassin commented:
![screenshot_2018-12-16_17-30-15](https://user-images.githubusercontent.com/4503202/50056131-4487a180-0158-11e9-8c6a-7960227d14c9.png)

Not sure if you consider this good style, but let me assure you, it's not, it's the worst I've seen in quite a while, actually. The first line of your commit is what is visible from e.g., `git log --oneline`, etc. Consider it the *heading* of your commit, it's the title that is supposed to show a reader what has been changed in a commit. Even pretty nonsense comments which are at least different from each other are easier to work with, they can at least 

Please see https://chris.beams.io/posts/git-commit/ or https://dev.to/pavlosisaris/git-commits-an-effective-style-guide-2kkn how to write better commit messages.

Also, please try to commit one change at a time, not multiple. Every time you need a bullet point list to describe your changes, you're probably trying to commit too much at once. This principle is also referred to as "atomic commits". See e.g., https://medium.com/@fagnerbrack/one-commit-one-change-3d10b10cebbf, for more information.

Bluesir9 commented:
@TheAssassin Point taken.

Bluesir9 commented on extractor/src/main/java/org/schabi/newpipe/extractor/utils/Base64.java:
```diff
@@ -1,3 +1,18 @@
+/*
```
Done.

Bluesir9 commented on extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/linkHandler/YoutubeSearchQueryHandlerFactory.java:
```diff
@@ -23,31 +107,129 @@ public static YoutubeSearchQueryHandlerFactory getInstance() {
+        return returnArray;
+    }
+
+    private List<Byte> convert(@Nonnull byte[] byteArray) {
```
Done.

theScrabi commented on third-party-licenses/AOSP-APACHE-LICENSE-V2.0:
```diff
@@ -0,0 +1,228 @@
+                            Apache License
```
I can't merge until this is fixed.

Bluesir9 replied:
Have removed the license.

Bluesir9 commented on extractor/src/main/java/org/schabi/newpipe/extractor/utils/Base64.java:
```diff
@@ -0,0 +1,725 @@
+ * Hence below implementation was picked from AOSP here ->
+ * https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/java/android/util/Base64.java
+ */
+public class Base64 {
```
Have removed the Base64 class I introduced and used the Apache commons library instead.

Bluesir9 commented on extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/linkHandler/YoutubeSearchQueryHandlerFactory.java:
```diff
@@ -2,19 +2,83 @@
+        Movie(FilterType.Content,(byte)0x04),
+        Show(FilterType.Content,(byte)0x05),
+
+        Hour(FilterType.Time,(byte)0x01),
```
I am not sure what you mean but this works out if the box. Albeit it looks kinda bad:

![Screenshot_1553365615](https://user-images.githubusercontent.com/5329780/54870295-a6484f00-4dca-11e9-86e8-65d8674fbf34.png)

theScrabi replied:
Yes this is the issue i ment, it is not possible to add or remove filters like key values. This messes up the UI as you see. The UI should be made ready to handle these filters.

Bluesir9 replied:
Maybe we could implement the filters at 2 levels? We would initially ask the user to select a `FilterType` and after they have chosen that, we can ask them the exact filter they wish to apply falling under that FilterType? 

So, clicking on the menu icon the first time, will load a list containing `Content, Time, Duration, Feature`. If the user chooses Duration, then we will show a second list containing `Short, Long`. And when the user chooses one of those options, then the search is triggered. Does that make sense?

TobiGr replied:
Let's finish this PR.  🎉 Thanks for being patient!

I am working on a proper design for the filters. This might take until next Sunday, because the new semesters is about to start and thus there are a bunch of other things to do. But I would not recommend to have endless lists. They come with bad UI/UX, because it is hard to get all the (important) information.

FYI: 
- [6 Best Practices forMobile App Search Filtering](https://www.raywenderlich.com/718-6-best-practices-for-mobile-app-search-filtering)
- [Filter and Sorting Best Practices on Mobile](https://medium.com/@thierrymeier/filtering-and-sorting-best-practices-on-mobile-61626449cec)
- [Search, Sort and Filter - Mobile Design Pattern Gallery](https://www.oreilly.com/library/view/mobile-design-pattern/9781449368586/ch04.html)
- [Collection of various screenshots](https://www.pinterest.de/darylhumphrey9/filter-mobile-ux/)
- [animations](https://uimovement.com/tag/filter/)

Bluesir9 replied:
@TobiGr Of course, take your time. I was just surprised and happy that it worked functionally out of the box.

TobiGr replied:
I put my UI ideas into [this issue](https://github.com/TeamNewPipe/NewPipe/issues/2251) in the app repo. I guess a UI discussion fits better into the other repo.

Bluesir9 commented:
While testing I found that the app is throwing an error in case of some filters(4K only for now). I will look into this and potentially push fixes for the same by tomorrow.

wb9688 commented:
Closing as it's based on YouTube's old site, introduces a new dependency which isn't neccessary, and there being no proper UI for it in NewPipe. Feel free to open a new PR in the future.
